### ISSUE TITLE
You can return an `$ElementType`, as long as you don’t assert that it’s of the right type

### ISSUE THREAD

**Issue by wchargin** [facebook/flow#6648]
Consider [the following example][1]:

```js
// @flow

type FooBox = {| +id: "FOO" |};
type BarBox = {| +id: "BAR" |};
type Box = FooBox | BarBox;

function good<T: Box>(x: T): $ElementType<T, "id"> {
  return x.id;
}

function wat<T: Box>(x: T): $ElementType<T, "id"> {
  return (x.id: $ElementType<T, "id">);  // error?!
}
```

[1]: https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AMVrAEKwAeoAvKAN4A+oA1AJYAmAXKAES4Dy3HoNAL4BuVJhyEAhgCdiZSrQYt2HQgEEASvyGj02UHIp4ChulNmlRiSAFcAdgGMUjWHdABzAswA8AFXZyAHwAFCTsvgCU7AAkAKLQWAC2WHYovuJ+ADScLByB1IigoNJYKDbSbiQAdCyigsi2js6uoPCSKH4BpCFhoJEx8UkpaRm+2Ry5+VSFxaXlbqE1bKBxCcmp6dhZOcx5EcJFIKBY0tKw0gD8AISI9UA

How can it be valid to return `e` if `e` is not of the return type?

Also, how is `x.id` not of type `$ElementType<typeof x, "id">`?

Using `$PropertyType` instead of `$ElementType` has the same result.

**Comment by goodmind** [comment]
```js
type FooBox = {| +id: "FOO" |};
type BarBox = {| +id: "BAR" |};

function wat<T: FooBox | BarBox>(x: T): $PropertyType<T, "id"> {
  return (x.id: $PropertyType<T, "id">);  // no error
}
```

This one works

**Comment by wchargin** [comment]
Wow! So it does: with either `$ElementType` or `$PropertyType`, inlining
the type alias causes the error to disappear.

This is [not the first time][1] that I’ve seen type aliases change the
behavior of Flow. Of all the things that I would expect to cause
confusing errors, type aliases are not one of them…

[1]: https://github.com/facebook/flow/issues/7302#issuecomment-451358792

**Comment by Bannerets** [comment]
The example in the OP seems to be fixed in v0.98.0

**Comment by wchargin** [comment]
Looks like it, yep. Thanks.
